### PR TITLE
[ao] Add new port libao

### PR DESCRIPTION
**Pull request by xiaoyifang**
**Describe the pull request**
https://github.com/microsoft/vcpkg/issues/21791
- #### What does your PR fix?  
add libao port

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
 windows, <No>

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

### CODE REVIEW

**Review by LilyWangLL — CHANGES_REQUESTED** [review]

**Comment by LilyWangLL** [comment]
Please run command `vcpkg x-add-version --all` then commit the changes.

**Comment by LilyWangLL** [comment]
There are error on Linux:
```
Include files should not be duplicated into the /debug/include directory. If this cannot be disabled in the project cmake, use
    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
```
Could you please fix this?

**Comment by xiaoyifang** [comment]
> There are error on Linux:
> 
> ```
> Include files should not be duplicated into the /debug/include directory. If this cannot be disabled in the project cmake, use
>     file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
> ```
> 
> Could you please fix this error?

I do not to fixed it in linux , can I just disabled the linux build

**Review by LilyWangLL — CHANGES_REQUESTED** [review]

**Comment by LilyWangLL** [comment]
Ao build failed on UWP with the following error:
```
     4>D:\buildtrees\ao\src\1.2.2-ef2f813c3f.clean\src\audio_out.c(654,26): error C4996: 'strcat': This function or variable may be unsafe. Consider using strcat_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details. [D:\buildtrees\ao\x64-uwp-dbg\ao.vcxproj]
     4>D:\buildtrees\ao\src\1.2.2-ef2f813c3f.clean\src\audio_out.c(655,17): error C4996: 'strcat': This function or variable may be unsafe. Consider using strcat_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details. [D:\buildtrees\ao\x64-uwp-dbg\ao.vcxproj]
```

**Comment by xiaoyifang** [comment]
> 

what about just disable it with `supports` field in vcpkg.json

**Review by LilyWangLL — APPROVED** [review]

File: ports/ao/0001-windows-build-patch.patch [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+-                              const char *filename,
+-                              int overwrite,
+-                              ao_sample_format *format,
+-                              ao_option *option);
+-
+-int                   ao_play(ao_device *device,
+-                              char *output_samples,
+-                              uint_32 num_bytes);
+-int                  ao_close(ao_device *device);
++LIBAO_DLL_EXPORT int ao_append_global_option(const char *key,
```
Comment (ras0219-msft) [inline ports/ao/0001-windows-build-patch.patch:66]: Adding individual DLL exports is unfortunately too much of a maintenance burden long term. If upstream does not offer that, we should instead mark the library as being static only on windows:

```
# portfile.cmake, at the top
if(VCPKG_TARGET_IS_WINDOWS)
    vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
endif()
```

File: ports/ao/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  OUT_SOURCE_PATH SOURCE_PATH
+  REPO xiph/libao
+  REF 1.2.2
+  SHA512  d2736d25b60862e7d7469611ce31b1df40a4366ab160e2ff1b46919ae91692d1596c8468e4f016303b306fc3ac1bddc7b727f535a362f403c3fe7c6532e9045a
+  HEAD_REF master
+  PATCHES ${PATCHES}
+)
+
+
+if(VCPKG_TARGET_IS_WINDOWS)
```
Comment (ras0219-msft) [inline ports/ao/portfile.cmake:15]: Did you try just using `vcpkg_configure_make` on Windows? We prefer to avoid having different buildsystems on different platforms because it tends to cause a bunch of maintenance issues long term.

File: ports/ao/vcpkg.json [hunk 3]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "ao",
```
Comment (ras0219-msft) [inline ports/ao/vcpkg.json:2]:
```suggestion
  "name": "libao",
```

It looks like the library is pretty universally called `libao` elsewhere [1], so we'd like to use that name here as well. You'll need to rename `ports/ao` -> `ports/libao`.

[1] https://repology.org/project/libao/versions

**Review by ras0219-msft — CHANGES_REQUESTED** [review]
Thanks for the PR!

Under normal circumstances we would not accept new libraries with such short names, however due to `libao`'s maturity and renown I believe it is acceptable to use `libao` in this case.

File: ports/ao/0001-windows-build-patch.patch (hunk 1 above)
Comment (xiaoyifang) [reply ports/ao/0001-windows-build-patch.patch]: but ,I really need a shared dll .
any other way to solve this.

File: ports/ao/vcpkg.json (hunk 3 above)
Comment (Neumann-A) [reply ports/ao/vcpkg.json]: > It looks like the library is pretty universally called `libao` elsewhere [1]

It calls itself just `ao`. https://github.com/xiph/libao/blob/20dc8ed9fa4605f5c25e7496ede42e8ba6468225/ao.pc.in#L8-L9
but is inconsistent about it : https://github.com/xiph/libao/blob/20dc8ed9fa4605f5c25e7496ede42e8ba6468225/ao.pc.in#L1

personally: the lib prefix is just a habit of linux to avoid collision with a binary called the same name. 
I really prefer the name from pc since it directly indicates usage: e.g.
`pkg-config --exists ao`

File: ports/ao/0001-windows-build-patch.patch (hunk 1 above)
Comment (xiaoyifang) [reply ports/ao/0001-windows-build-patch.patch]: and the method signature has not changed for about 10 years.
the project has not under active development.

Comment (ras0219-msft) [reply ports/ao/0001-windows-build-patch.patch]: If these are C files, then using a `.def` file may be acceptable. It lists out the function names to be exported and doesn't require extremely large patches.

Example: https://github.com/microsoft/vcpkg/blob/master/ports/libass/libass.def

File: ports/ao/vcpkg.json (hunk 3 above)
Comment (ras0219-msft) [reply ports/ao/vcpkg.json]: I see:
1. https://github.com/xiph/libao
2. Website url is https://www.xiph.org/ao/
2. Official description: > Libao is a cross-platform audio library
3. Official source code tarball is [libao-1.2.0.tar.gz](http://downloads.xiph.org/releases/ao/libao-1.2.0.tar.gz)
4. pkg-config is `ao`
5. The vast majority of other linuxes use `libao` (except GUIX uses `ao` https://guix.gnu.org/packages/ao-1.2.2-5-g20dc8ed/)
6. Brew uses https://formulae.brew.sh/formula/libao

I agree and understand that linuxes are motivated to use the lib prefix to avoid conflicts, but I think it promotes portability the most to use the common package name. It is very unfortunate that it disagrees with pkg-config though.

File: ports/ao/0001-windows-build-patch.patch (hunk 1 above)
Comment (xiaoyifang) [reply ports/ao/0001-windows-build-patch.patch]: after off-work ,I will change it to def

File: ports/ao/vcpkg.json (hunk 3 above)
Comment (xiaoyifang) [reply ports/ao/vcpkg.json]: what about a compromise ,project name libao ,generate libarary ao.lib on windows.

File: ports/libao/0001-windows-build-patch.patch [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+ 
++#define LIBAO_DLL_EXPORT
++#ifdef _WIN32
++#undef LIBAO_DLL_EXPORT
++#ifdef _WINDLL
++#define LIBAO_DLL_EXPORT __declspec(dllexport)
++#else
++#define LIBAO_DLL_EXPORT __declspec(dllimport)
++#endif
++#endif
```
Comment (Neumann-A) [inline ports/libao/0001-windows-build-patch.patch:32]: Do you still need this after your switch to `vcpkg_configure_make`?
Libtool typically does and export of all symbols and `dllimport` is implied unless it is a data symbol (global variable.)

File: ports/libao/portfile.cmake [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  vcpkg_configure_make(
+    SOURCE_PATH ${SOURCE_PATH}
+    AUTOCONFIG
+    OPTIONS --disable-binaries # no example programs (require libogg)
+  )
+  vcpkg_install_make()
+
+  file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/share")
+  file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
+endif()
```
Comment (Neumann-A) [inline ports/libao/portfile.cmake:33]: What is the error on windows ?

File: ports/ao/vcpkg.json (hunk 3 above)
Comment (Neumann-A) [reply ports/ao/vcpkg.json]: `vcpkg_configure_make` should do the right thing about output names. 

> I agree and understand that linuxes are motivated to use the lib prefix to avoid conflicts, but I think it promotes portability the most to use the common package name.

I just find `lib<portname>` always hard to parse in between the other `lib<name>` stuff. Especially if `<portname>` is short. If we use a vote system I would say it is 3(github|description|most-other-package-name):2(pkg-config|website) for `libao`.

Personally I think vcpkg should have a consistent setting here and following upstream pkg-config seems to be our best defined bet. Especially since there are ports doing it correctly in pkg-config e.g. `libpng`. (I mean if we go with the linux package name approach `zlib` should have been called `libz`)

File: ports/libao/0001-windows-build-patch.patch (hunk 4 above)
Comment (xiaoyifang) [reply ports/libao/0001-windows-build-patch.patch]: removed .

Comment (xiaoyifang) [reply ports/libao/0001-windows-build-patch.patch]: removed

File: ports/libao/portfile.cmake (hunk 5 above)
Comment (xiaoyifang) [reply ports/libao/portfile.cmake:33]: ao_wmm.obj : error LNK2001: unresolved external symbol KSDATAFORMAT_SUBTYPE_PCM
.libs\ao-4.dll : fatal error LNK1120: 1 unresolved external
[build-x64-windows-rel-rel-out.zip](https://github.com/microsoft/vcpkg/files/8080097/build-x64-windows-rel-rel-out.zip)
[build-x64-windows-rel-rel-err.zip](https://github.com/microsoft/vcpkg/files/8080098/build-x64-windows-rel-rel-err.zip)
s

File: ports/ao/vcpkg.json (hunk 3 above)
Comment (Neumann-A) [reply ports/ao/vcpkg.json]: to further add to this. vcpkg could just add both ports `ao` and `libao` and make one of them a stub.

File: ports/libao/portfile.cmake (hunk 5 above)
Comment (Neumann-A) [reply ports/libao/portfile.cmake:33]: Seems to be a windows symbol. (https://docs.microsoft.com/en-us/windows/win32/api/mmreg/ns-mmreg-waveformatextensible)
So maybe just an additional library needs to be linked which fixes the build?

Comment (Neumann-A) [reply ports/libao/portfile.cmake:33]: that being said I see your cmake code linking  `winmm ksuser`. So the fix is probably to add those to the make build on windows and it will just work.

File: ports/libao/0001-windows-build-patch.patch [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+@@ -0,0 +1,9 @@
++#pragma once
++#include <stdint.h>
++
++typedef uint8_t  uint_8;
++typedef uint16_t uint_16;
++typedef uint32_t uint_32;
++typedef int8_t   sint_8;
++typedef int16_t  sint_16;
++typedef int32_t  sint_32;
```
Comment (Neumann-A) [inline ports/libao/0001-windows-build-patch.patch:117]: in the make build it is likely you don't need these anymore?

File: ports/libao/portfile.cmake (hunk 5 above)
Comment (Neumann-A) [reply ports/libao/portfile.cmake:33]: If you need a quick fix `set(ENV{LIBS} "winmm  ksuser")` in the portfile should do the trick.

Comment (Neumann-A) [reply ports/libao/portfile.cmake:33]: Looking at the configure script you might also want to pass `--enable-wmm`

File: ports/libao/0001-windows-build-patch.patch (hunk 6 above)
Comment (xiaoyifang) [reply ports/libao/0001-windows-build-patch.patch:117]: if vcpkg_make works ,this can be remove also.

File: ports/ao/vcpkg.json (hunk 3 above)
Comment (Neumann-A) [reply ports/ao/vcpkg.json]: let @ras0219-msft decide before taking action ;)

File: ports/libao/portfile.cmake (hunk 5 above)
Comment (xiaoyifang) [reply ports/libao/portfile.cmake:33]: same error
image

[build-x64-windows-rel-rel-out (2).zip](https://github.com/microsoft/vcpkg/files/8080264/build-x64-windows-rel-rel-out.2.zip)

Comment (Neumann-A) [reply ports/libao/portfile.cmake:33]: `libtool: link:  compile cl.exe -o .libs\\ao-4.dll  .libs/audio_out.obj .libs/config.obj .libs/ao_null.obj .libs/ao_wav.obj .libs/ao_au.obj .libs/ao_raw.obj .libs/ao_aixs.obj .libs/ao_wmm.obj   -O -O2 -Oi    -LD:/vcpkg/installed/x64-windows-rel/lib -LD:/vcpkg/installed/x64-windows-rel/lib/manual-link -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -lcomdlg32 -ladvapi32 -lwinmm "@.libs\\ao-4.dll.exp" -Wl,-DLL,-IMPLIB:".libs\\ao.lib"`

Not linking `ksuser.lib` so `LIBS` is ignored for some reason. (maybe the project is overwriting it.)

ah maybe the solution is to use `-l` prefix so use `set(ENV{LIBS} "-lksuser")` so libtool can correctly detect it.

**Comment by xiaoyifang** [comment]
normally ,when will this be merge?

**Comment by xiaoyifang** [comment]
can not be searched in the following website
https://vcpkg.io/en/packages.html
